File: session_rust/src/point.rs
use crate::Color;
use serde::{ser::Serialize as SerTrait, Deserialize, Serialize};
use std::fmt;
use uuid::Uuid;

/// A 3D point with visual properties and JSON serialization support.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename = "Point")]
pub struct Point {
    pub guid: String,      // Unique identifier
    pub name: String,      // Name of the point
    pub x: f32,            // X coordinate
    pub y: f32,            // Y coordinate
    pub z: f32,            // Z coordinate
    pub width: f32,        // Width of the point
    pub pointcolor: Color, // Color of the point
}

impl Default for Point {
    fn default() -> Self {
        Self {
            x: 0.0,
            y: 0.0,
            z: 0.0,
            guid: Uuid::new_v4().to_string(),
            name: "my_point".to_string(),
            pointcolor: Color::white(),
            width: 1.0,
        }
    }
}

impl Point {
    /// Creates a new Point with specified coordinates.
    pub fn new(x: f32, y: f32, z: f32) -> Self {
        Self {
            x,
            y,
            z,
            ..Default::default()
        }
    }

    ///////////////////////////////////////////////////////////////////////////////////////////
    // JSON
    ///////////////////////////////////////////////////////////////////////////////////////////

    /// Serializes the Point to a JSON string.
    pub fn to_json_data(&self) -> Result<String, Box<dyn std::error::Error>> {
        let mut buf = Vec::new();
        let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, formatter);
        SerTrait::serialize(self, &mut ser)?;
        Ok(String::from_utf8(buf)?)
    }

    /// Deserializes a Point from a JSON string.
    pub fn from_json_data(json_data: &str) -> Result<Self, Box<dyn std::error::Error>> {
        Ok(serde_json::from_str(json_data)?)
    }

    /// Serializes the Point to a JSON file.
    pub fn to_json(&self, filepath: &str) -> Result<(), Box<dyn std::error::Error>> {
        let json = self.to_json_data()?;
        std::fs::write(filepath, json)?;
        Ok(())
    }

    /// Deserializes a Point from a JSON file.
    pub fn from_json(filepath: &str) -> Result<Self, Box<dyn std::error::Error>> {
        let json = std::fs::read_to_string(filepath)?;
        Self::from_json_data(&json)
    }
}

impl fmt::Display for Point {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Point({}, {}, {}, {}, {}, {}, {})",
            self.x, self.y, self.z, self.guid, self.name, self.pointcolor, self.width
        )
    }
}

File: session_py/src/session_py/point.py
import uuid
import json
from .color import Color


class Point:
    """A 3D point with visual properties.

    Parameters
    ----------
    x : float, optional
        X coordinate. Defaults to 0.0.
    y : float, optional
        Y coordinate. Defaults to 0.0.
    z : float, optional
        Z coordinate. Defaults to 0.0.

    Attributes
    ----------
    name : str
        The name of the point.
    guid : str
        The unique identifier of the point.
    x : float
        The X coordinate of the point.
    y : float
        The Y coordinate of the point.
    z : float
        The Z coordinate of the point.
    pointcolor : :class:`Color`
        The color of the point.
    width : float
        The width of the point for display.

    """

    def __init__(self, x=0.0, y=0.0, z=0.0):
        self.guid = str(uuid.uuid4())
        self.name = "my_point"
        self.x = x
        self.y = y
        self.z = z
        self.width = 1.0
        self.pointcolor = Color.white()

    def __str__(self):
        return f"Point({self.x}, {self.y}, {self.z})"

    def __repr__(self):
        return f"Point({self.x}, {self.y}, {self.z}, {self.guid}, {self.name}, {self.pointcolor}, {self.width})"

    def __eq__(self, other):
        return (
            self.name == other.name
            and round(self.x, 6) == round(other.x, 6)
            and round(self.y, 6) == round(other.y, 6)
            and round(self.z, 6) == round(other.z, 6)
            and round(self.width, 6) == round(other.width, 6)
            and self.pointcolor == other.pointcolor
        )

    ###########################################################################################
    # JSON
    ###########################################################################################

    def to_json_data(self) -> dict:
        """Convert the Point to a JSON-serializable dictionary.

        Returns
        -------
        dict
            Dictionary containing the Point data in JSON format.

        """
        return {
            "type": "Point",
            "guid": self.guid,
            "name": self.name,
            "x": self.x,
            "y": self.y,
            "z": self.z,
            "width": self.width,
            "pointcolor": self.pointcolor.to_json_data(),
        }

    @classmethod
    def from_json_data(cls, data: dict) -> "Point":
        """Create a Point from JSON data dictionary.

        Parameters
        ----------
        data : dict
            Dictionary containing point data from JSON.

        Returns
        -------
        :class:`Point`
            Point instance created from the JSON data.

        """
        point = cls(data["x"], data["y"], data["z"])
        point.guid = data["guid"]
        point.name = data["name"]
        point.width = data["width"]
        point.pointcolor = Color.from_json_data(data["pointcolor"])
        return point

    def to_json(self, filepath: str) -> None:
        """Serialize the Point to a JSON file.

        Parameters
        ----------
        filepath : str
            Path to the output JSON file.

        """
        with open(filepath, "w") as f:
            json.dump(self.to_json_data(), f, indent=4)

    @classmethod
    def from_json(cls, filepath: str) -> "Point":
        """Deserialize a Point from a JSON file.

        Parameters
        ----------
        filepath : str
            Path to the JSON file to load.

        Returns
        -------
        :class:`Point`
            Point instance loaded from the file.

        """
        with open(filepath, "r") as f:
            data = json.load(f)
            return cls.from_json_data(data)

    ###########################################################################################
    # Details
    ###########################################################################################
